service_test.go

```diff
@@ -9,40 +9,90 @@ import (
-	}
-	tests := []struct {
-		name    string
-		args    args
-		wantErr bool
+	})
+
+	var httpBuilderAllErrors = errors.New("name is required\n" +
+		"provided routes slice was empty\n" +
+		"provided middlewares slice was empty\n" +
+		"alive check func provided was nil\n" +
+		"ready check func provided was nil\n" +
+		"provided components slice was empty\n" +
+		"provided SIGHUP handler was nil\n")
+
+	tests := map[string]struct {
+		name          string
+		version       string
+		cps           []Component
+		routes        []phttp.Route
+		middlewares   []phttp.MiddlewareFunc
+		acf           phttp.AliveCheckFunc
+		rcf           phttp.ReadyCheckFunc
+		sighupHandler func()
+		wantErr       error
 	}{
-		{"success", args{name: "test", opt: []OptionFunc{Routes([]phttp.Route{route}), Middlewares(middleware)}}, false},
-		{"failed empty middlewares", args{name: "test", opt: []OptionFunc{Routes([]phttp.Route{route}), Middlewares([]phttp.MiddlewareFunc{}...)}}, true},
-		{"failed missing name", args{name: "", opt: []OptionFunc{Routes([]phttp.Route{route})}}, true},
-		{"failed missing routes", args{name: "test", opt: []OptionFunc{Routes([]phttp.Route{})}}, true},
+		"success": {
+			name:          "test",
+			version:       "dev",
+			cps:           []Component{&testComponent{}},
+			routes:        []phttp.Route{route},
+			middlewares:   []phttp.MiddlewareFunc{middleware},
+			acf:           phttp.DefaultAliveCheck,
+			rcf:           phttp.DefaultReadyCheck,
+			sighupHandler: func() { log.Info("SIGHUP received: nothing setup") },
+			wantErr:       nil,
+		},
+		"nil inputs steps": {
+			name:          "",
+			version:       "",
+			cps:           nil,
+			routes:        nil,
+			middlewares:   nil,
+			acf:           nil,
+			rcf:           nil,
+			sighupHandler: nil,
+			wantErr:       httpBuilderAllErrors,
+		},
+		"error in all builder steps": {
+			name:          "",
+			version:       "",
+			cps:           []Component{},
+			routes:        []phttp.Route{},
+			middlewares:   []phttp.MiddlewareFunc{},
+			acf:           nil,
+			rcf:           nil,
+			sighupHandler: nil,
+			wantErr:       httpBuilderAllErrors,
+		},
 	}
-	for _, tt := range tests {
-		t.Run(tt.name, func(t *testing.T) {
-			got, err := New(tt.args.name, "", tt.args.opt...)
-			if tt.wantErr {
-				assert.Error(t, err)
-				assert.Nil(t, got)
+
+	for name, tt := range tests {
+		t.Run(name, func(t *testing.T) {
+			gotService, gotErr := NewBuilder(tt.name, tt.version).
+				WithRoutes(tt.routes).
+				WithMiddlewares(tt.middlewares...).
+				WithAliveCheck(tt.acf).
+				WithReadyCheck(tt.rcf).
+				WithComponents(tt.cps...).
+				WithSIGHUP(tt.sighupHandler).Build()
+
+			if tt.wantErr != nil {
+				assert.Equal(t, tt.wantErr.Error(), gotErr.Error())
+				assert.Nil(t, gotService)
 			} else {
-				assert.NoError(t, err)
-				assert.NotNil(t, got)
+				assert.Nil(t, gotErr)
+				assert.NotNil(t, gotService)
+				assert.IsType(t, &Service{}, gotService)
```

Comment:
we do not actually test that the builder created the service we wanted. We should check that if we add routes the service will contain them, and all the rest of the setup of course.

Reply:
@mantzas I'm on the fence on what's the best way to approach this.

Comparison of `func` types and structs containing them, such as `routes`, is not allowed. Ideally, I'd like to avoid hacks such as printing out and comparing memory addresses, or `reflect.Value.Pointer()`.

The testify/assert package uses `reflect.DeepEqual` in the background, which also [cannot be used](https://golang.org/src/reflect/deepequal.go#L171) for this purpose. Maybe we could somehow test the *functionality*?

Reply:
No need for that, but there are things that we can test since the service contains components, middleware, etc and we can always check that these funcs are not nil...

Reply:
I've added some more assertions, feel free to check them out and see if they're overkill or if they make sense.